### PR TITLE
67/gql filters

### PR DESCRIPTION
**Pull request by MattDodsonEnglish**
- Add reference and guide
- refactor overview page a bit
Part of work on #67

### CODE REVIEW

File: content/how-to/gql/query-filters.md [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+        "_createdOn": "2023-12-24T16:50:45Z",
+        "_modifiedOn": "2024-01-23T20:06:30Z"
+      },
+      {
+        "id": "AN-19670-Equipment-2",
+        "_createdOn": "2023-12-24T18:16:35Z",
+        "_modifiedOn": "2024-01-23T20:06:35Z"
+      },
+      // more items
+      ]
```
Comment (MattDodsonEnglish) [inline content/how-to/gql/query-filters.md:77]:
```suggestion
      ]
    }
 }
```

File: content/how-to/gql/query-filters.md [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+    }
+  },
+}
+```
+{{% /tab %}}
+{{< /tabs >}}
+
+## Combine filters with `and`, `or`, `not`
+
+To filter by multiple properties, use the `and`, `or` and `not`, operators.
```
Comment (MattDodsonEnglish) [inline content/how-to/gql/query-filters.md:149]:
```suggestion
To filter by multiple properties, use the `and`, `or`, and `not`, operators.
```

File: content/how-to/gql/query-filters.md [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+GraphQL syntax uses [infix notation](https://en.wikipedia.org/wiki/Infix_notation), so: "a and b" is `a, and: { b }`, “a or b or c” is `a, or: { b, or: c }`, and “not” is a prefix (`not:`).
+
+### this `and` that property
+
+The `and` operator filters for objects that include all specified properties.
+
+For example, this query returns equipment objects that match two properties:
+- `effectiveStart` must be the 1st and the 10th of January, 2024
```
Comment (MattDodsonEnglish) [inline content/how-to/gql/query-filters.md:157]:
```suggestion
- The `effectiveStart` must be the 1st and the 10th of January, 2024.
```

File: content/how-to/gql/query-filters.md [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+- It must have a non-null `nextVersion`
```
Comment (MattDodsonEnglish) [inline content/how-to/gql/query-filters.md:158]:
```suggestion
- It must have a non-null `nextVersion`.
```

File: content/how-to/gql/query-filters.md [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+    effectiveStart
+    id
+    nextVersion
+  }
+}
+```
+
+{{< notice "note" >}}
+
+This is shorter filter syntax is equivalent to `and: {has: nextVersion}`.
```
Comment (MattDodsonEnglish) [inline content/how-to/gql/query-filters.md:184]:
```suggestion
This preceding filter syntax is a shorter equivalent to `and: {has: nextVersion}`.
```

File: content/how-to/gql/query-filters.md [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+
+To modify this to include both objects within the range and objects that do not have a `nextVersion`, use `or` with `not`:
+
+```graphql
+or: { not: {has: nextVersion} }
+```
+
+### A list of filters
+
+The `and` and `or` accept lists of filters.
```
Comment (MattDodsonEnglish) [inline content/how-to/gql/query-filters.md:231]:
```suggestion
The `and` and `or` operators accept lists of filters.
```

File: content/how-to/gql/query-filters.md [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+For example, this query filters for equipment objects whose `id` is matches `A`, `B`, or `C`:
```
Comment (MattDodsonEnglish) [inline content/how-to/gql/query-filters.md:232]:
```suggestion
For example, this query filters for equipment objects whose `id` matches `A`, `B`, or `C`:
```

File: content/how-to/gql/query-filters.md [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+## Use directives
+
+Rhize offers [_query directives_](https://the-guild.dev/graphql/tools/docs/schema-directives#what-about-query-directives), special instructions about how to look up and return values in a query.
+These directives can extend your filtering to look at nested properties or to conditionally display a field.
+
+All directives begin with the `@` sign.
+
+### Cascade
+
+The `@cascade`  directive filters certain nodes within a query.
```
Comment (MattDodsonEnglish) [inline content/how-to/gql/query-filters.md:253]:
```suggestion
The `@cascade`  directive filters for certain nodes within a query.
```

File: content/how-to/gql/query-filters.md [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+Use it to filter requested resources by a nested sub-property, similar to a `WHERE` clause in SQL.
+
+{{< notice "caution" >}}
+
+`@cascade` is not as performant as flatter queries.
+Consider using it only after you've exhausted other query structures to return the data you want.
+
+{{< /notice >}}
+
+For example, this query [filters]({{< relref "call-the-graphql-api#filter" >}}) for job responses with an ID of `12341`, and then filters that set for only the items that have a `data.properyLabel` field with a value of `INSTANCE ID`.
```
Comment (MattDodsonEnglish) [inline content/how-to/gql/query-filters.md:263]:
```suggestion
For example, this query filters for job responses with an ID of `12341`, and then filters that set for only the items that have a `data.properyLabel` field with a value of `INSTANCE ID`.
```

File: content/how-to/gql/query-filters.md [hunk 10]
```diff
@@ -0,0 +1,337 @@
...
+}
+
+```
+{{% /tab %}}
+{{< /tabs >}}
+
+### Include
+
+The `@include` directive returns a field only if its variable is `true`.
+It is the converse of [`@skip`](#skip), which returns a field if its variable is `false`.
```
Comment (MattDodsonEnglish) [inline content/how-to/gql/query-filters.md:301]:
```suggestion
```

File: content/reference/gql-filters.md [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+## Geolocation filters
+
+Geolocation filters filter by geographic coordinates.
+They return matches within the specified [GeoJSON polygon](https://datatracker.ietf.org/doc/html/rfc7946#section-3.1.6).
+
+| Filter     | Description                                     |
+|------------|-------------------------------------------------|
+| near       | Within the specified `distance` from the polyon |
+| within     | In the polygon coordinates                      |
+| Intersects | In the intersection of two polygons                                                |
```
Comment (MattDodsonEnglish) [inline content/reference/gql-filters.md:70]:
```suggestion
| Intersects | In the intersection of two polygons                     |

```

File: content/how-to/gql/filter.md [hunk 12]
```diff
@@ -0,0 +1,344 @@
...
+weight: 210
+menu:
+  main:
+    parent: how-to-query
+    identifier:
+---
+
+When a query or mutation involves many items, it's often helpful to filter the response to some subset.
+Rhize has many filters to make your queries more precise, bringing more meaningful results with less need for secondary processing.
+
```
Comment (MattDodsonEnglish) [inline content/how-to/gql/filter.md:14]:
```suggestion

_Filters_  limit an operation to a subset of resources.
You can use filters to make operations more precise, remove unneeded items from a payload, and reduce the need for secondary processing.

To use a filter, specify it in the operation's argument.
Most fields in an object can serve as a filter.
```

File: content/how-to/gql/mutate.md [hunk 13]
```diff
@@ -71,9 +71,23 @@ mutation AddEquipment($input: [AddEquipmentInput!]!) {
 ```
 {{% /tab %}}{{< /tabs >}}
 
-### `update`
+### Upsert
+
+Many `add` operations support _upserting_, which _update_ or _insert_ (create).
+That is, if the object already exists, the operation will update it with the additional fields.
```
Comment (MattDodsonEnglish) [inline content/how-to/gql/mutate.md:77]:
```suggestion
That is, if the object already exists, the operation will update it with the additional fields.
If the object doesn't exist, the operation will create it.
```

File: content/how-to/gql/mutate.md [hunk 14]
```diff
@@ -71,9 +71,23 @@ mutation AddEquipment($input: [AddEquipmentInput!]!) {
...
+
+Besides general UX convenience, upsert is useful when data comes from multiple sources and in no guaranteed order, or when data comes from the message broker.
```
Comment (MattDodsonEnglish) [inline content/how-to/gql/mutate.md:79]:
```suggestion
Besides general UX convenience, upsert is useful when data comes from multiple sources and in no guaranteed order, like from multiple streams from the message broker.
```

File: content/how-to/gql/mutate.md [hunk 15]
```diff
@@ -150,4 +165,81 @@ mutation deleteUoM($filter: UnitOfMeasureFilter!){
...
+    "equipment": {
+      "id": "widget_maker_1",
+      "label": "Widget maker 1"
+
+    }
+  }
+}
+```
+
+You can confirm the record and its nested property exists with a `get` query.
```
Comment (MattDodsonEnglish) [inline content/how-to/gql/mutate.md:204]:
```suggestion
You can confirm that the record and its nested property exists with a `get` query.
```

File: content/reference/gql-types.md [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+### Object type
+
+The preceding basic types form the building blocks for Rhize's manufacturing object schema, with data models corresponding to ISA-95.
+
+Each object is made of manufacturing-specific fields of one of the basic types.
+For example, the `materialActual`  object has basic fields, including:
+- `description`, a `String`.
+- `effectiveEnd`, a `DateTime`
+
+The `materialActual` also has complex fields describing associated manufacturing objects, for example,
```
Comment (MattDodsonEnglish) [inline content/reference/gql-types.md:53]:
```suggestion
The `materialActual` also has complex fields describing associated manufacturing objects.
For example, its fields include
```

File: content/reference/gql-types.md [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+an array of associated `MaterialLot` objects, the `MaterialDefinition` object, and so on.
```
Comment (MattDodsonEnglish) [inline content/reference/gql-types.md:54]:
```suggestion
the array of associated `MaterialLot` objects, the `MaterialDefinition` object, and so on.
```

File: content/reference/gql-types.md [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+Properties of the type `Enum` can be filtered by the following:
+ - `lt`
+ - `le`
+ - `eq`
+ - `in`
+ - `between`
+ - `ge`
+ - `gt`
+
+ Each keyword has the same behavior as described in [string filters](#string-filters).
```
Comment (MattDodsonEnglish) [inline content/reference/gql-types.md:101]:
```suggestion
Each keyword has the same behavior as described in [string filters](#string-filters).
```

File: content/reference/gql-types.md [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+### Basic types
+
+Every manufacturing object in the Rhize database is made of fields that are of one the following basic types.
+In official GraphQL terminology, these types are called [_scalar types_](https://graphql.org/learn/schema/#scalar-types).
+
+- `String`
+- `Int`
+- `Float`
+- `Enum`
+- `Boolean`
```
Comment (Jarrah-libremfg) [inline content/reference/gql-types.md:38]: Should we define these? I could see most people getting string and maybe int, but float and enum especially are not terms most people are familiar with, though not sure about people from a manufacturing background.

File: content/reference/gql-types.md [hunk 20]
```diff
@@ -0,0 +1,122 @@
...
+- `id`: A string representing a unique object within a defined [object type](#object-type).
+- `iid`: The object's unique address in the database. For example, `0xf9b49`.
+- `DateTime`: A timestamp in [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) format.
```
Comment (Jarrah-libremfg) [inline content/reference/gql-types.md:41]: Copied from slack:

Our libre-schema DateTime implementation just directly calls time.Parse(), so I'd say most of the dates that aren't stored as strings and explicitly defined as ISO8601Foo would be RFC3339.

File: content/reference/gql-types.md [hunk 21]
```diff
@@ -0,0 +1,122 @@
...
+
+The `materialActual` also has complex fields describing associated manufacturing objects.
+For example, its fields include
+the array of associated `MaterialLot` objects, the `MaterialDefinition` object, and so on.
+All objects in the database have relationships to other objects.
+
+## Scalar filters
+
+Every field that has a basic type can be used as a filter.
+The filters that are available depend on the data type, but the behavior of the `String` filters corresponds closely to `DateTime` and `Int` filters.
```
Comment (Jarrah-libremfg) [inline content/reference/gql-types.md:61]: This doesn't seem true based on a quick attempt. The filter on `queryWorkflowSpecification` allows me to filter an `id` by the methods listed under "String filters", but not for example the `_modifiedBy` field. 

My understanding is that these filters are defined as `@search(by: ["foo"])` directives in the schema (and potentially also by `@id`).

To see this effect in action, compare the filter for `workflowSpecification` to that of `Rule`. The former does not have `@search` on `_modifiedBy` and thus can't be filtered by that field. The latter does.

filters

File: content/reference/gql-types.md [hunk 22]
```diff
@@ -0,0 +1,122 @@
...
+| `eq`                           | Equals (exact match)                                      | `(filter: {id: {eq: "match"}})`                  |
+| `in`                           | From this match of arrays                                 | `(filter: {id: {in: ["dough", "cookie_unit"]}})` |
+| `lt`,`gt`,`le`, `ge` `between` | Less than, greater than, or between a lexicographic range | `(filter: {id: {lt: "M"}`                        |
+| `regexp`                       | A regular expression match using [`RE2` syntax](https://github.com/google/re2/wiki/Syntax/)                | `(filter: {id: {regexp: "/hello/i"}})`           |
+| `anyoftext`                    | A match for any entered strings, separated by spaces      | `(filter: {id: {anyoftext: "100 MD"}})`          |
+| `alloftext`                    | A match for all entered strings                           | `(filter: {id {alloftext: "MD"}})`               |
+
+### Integers, floats, DateTimes
+
+Properties that have a type of `Int`, `Float`, or `DateTime` can be filtered by the following keywords.
```
Comment (Jarrah-libremfg) [inline content/reference/gql-types.md:78]: The same as above applies here, try filtering the `rules` of a `workflowSpecificaiton` based on its `delayTime` for example.

File: content/reference/gql-types.md [hunk 23]
```diff
@@ -0,0 +1,122 @@
...
+### Geolocation filters
+
+Geolocation filters return objects within specified geographic coordinates.
+They return matches within the specified [GeoJSON polygon](https://datatracker.ietf.org/doc/html/rfc7946#section-3.1.6).
+
+| Filter     | Description                                      |
+|------------|--------------------------------------------------|
+| near       | Within the specified `distance` from the polygon |
+| within     | In the polygon coordinates                       |
+| Intersects | In the intersection of two polygons              |
```
Comment (Jarrah-libremfg) [inline content/reference/gql-types.md:117]: There is also Contains here.

I'd note that at the moment we don't seem to have any `Point` or `Polygon` (the types that implement this) that have `@search` defined.

File: content/reference/gql-types.md (hunk 19 above)
Comment (MattDodsonEnglish) [inline content/reference/gql-types.md:38]:
```suggestion
- `String`: A sequence of characters. For example, `machine_2`
- `Int`: An integer number.  For example, `2`.
- `Float`: A number that includes a fraction. For example, `2.25`.
- `Boolean`: A field whose value is either `true` or `false`.
- `Enum`: A field whose values are restricted to a defined set. For example, `versionState` might be one of `ACTIVE`, `APPROVED`, `FOR_REVIEW`, `DRAFT`, or `DEPRECATED`

```

File: content/reference/gql-types.md [hunk 24]
```diff
@@ -0,0 +1,123 @@
...
+
+- `String`: A sequence of characters. For example, `machine_2`
+- `Int`: An integer number.  For example, `2`.
+- `Float`: A number that includes a fraction. For example, `2.25`.
+- `Boolean`: A field whose value is either `true` or `false`.
+- `Enum`: A field whose values are restricted to a defined set. For example, `versionState` might be one of `ACTIVE`, `APPROVED`, `FOR_REVIEW`, `DRAFT`, or `DEPRECATED`
+
+- `id`: A string representing a unique object within a defined [object type](#object-type).
+- `iid`: The object's unique address in the database. For example, `0xf9b49`.
+- `DateTime`: A timestamp in [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) format.
```
Comment (MattDodsonEnglish) [inline content/reference/gql-types.md:42]:
```suggestion
- `DateTime`: A timestamp in [RFC 3339](https://datatracker.ietf.org/doc/html/rfc3339) format.
```

File: content/reference/gql-types.md (hunk 21 above)
Comment (MattDodsonEnglish) [reply content/reference/gql-types.md:62]: Whoa! Good catch. That would have been a _huge_ inaccuracy. I think I just had the "bad luck" to do all my testing on objects with lots of filters.

File: content/reference/gql-types.md [hunk 25]
```diff
@@ -0,0 +1,123 @@
...
+
+### Boolean filters
+
+Boolean filters can be either `true` or `false`.
+
+### Geolocation filters
+
+Geolocation filters return objects within specified geographic coordinates.
+They return matches within the specified [GeoJSON polygon](https://datatracker.ietf.org/doc/html/rfc7946#section-3.1.6).
+
```
Comment (MattDodsonEnglish) [inline content/reference/gql-types.md:113]:
```suggestion

If a geolocation field can act as a filter, then the filter can work in one of the following behaviors:
```

File: content/reference/gql-types.md [hunk 26]
```diff
@@ -0,0 +1,124 @@
...
+| `in`                           | From this match of arrays                                 | `(filter: {id: {in: ["dough", "cookie_unit"]}})` |
+| `lt`,`gt`,`le`, `ge` `between` | Less than, greater than, or between a lexicographic range | `(filter: {id: {lt: "M"}`                        |
+| `regexp`                       | A regular expression match using [`RE2` syntax](https://github.com/google/re2/wiki/Syntax/)                | `(filter: {id: {regexp: "/hello/i"}})`           |
+| `anyoftext`                    | A match for any entered strings, separated by spaces      | `(filter: {id: {anyoftext: "100 MD"}})`          |
+| `alloftext`                    | A match for all entered strings                           | `(filter: {id {alloftext: "MD"}})`               |
+
+### Integers, floats, DateTimes
+
+Properties that have a type of `Int`, `Float`, or `DateTime` can be filtered by the following keywords.
+Note that all date times are in the [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) format.
```
Comment (MattDodsonEnglish) [inline content/reference/gql-types.md:80]:
```suggestion
```

File: content/reference/gql-types.md [hunk 27]
```diff
@@ -0,0 +1,124 @@
...
+ - `lt`
+ - `le`
+ - `eq`
+ - `in`
+ - `between`
+ - `ge`
+ - `gt`
+
+Each keyword has the same behavior as described in [string filters](#string-filters), only they operate on numerical rather than lexicographic values.
+
```
Comment (MattDodsonEnglish) [inline content/reference/gql-types.md:91]: {{< notice "note" >}}
While the `dateTime` type uses the RFC 3339 format, some string fields may use the [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) format. This depends on the object and customer requirement. For these fields, the string filters work as chronological filters too.
{{< /notice >}}

File: content/reference/gql-types.md [hunk 28]
```diff
@@ -0,0 +1,123 @@
...
```
Comment (MattDodsonEnglish) [inline content/reference/gql-types.md:90]:
```suggestion
{{< notice "note" >}}
While the dateTime type uses the RFC 3339 format, some string fields may use the [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) format. This depends on the object and customer requirement. For these fields, the string filters work as chronological filters too.
{{< /notice >}}
```

**Comment by MattDodsonEnglish** [comment]
Thanks for the review, @Jarrah-libremfg . I've added all your suggestions, and fixed the inaccuracies you pointed out.

File: content/reference/gql-types.md [hunk 29]
```diff
@@ -0,0 +1,125 @@
...
+ - `le`
+ - `eq`
+ - `in`
+ - `between`
+ - `ge`
+ - `gt`
+
+Each keyword has the same behavior as described in [string filters](#string-filters), only they operate on numerical rather than lexicographic values.
+{{< notice "note" >}}
+While the dateTime type uses the RFC 3339 format, some string fields may use the [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) format. This depends on the object and customer requirement. For these fields, the string filters work as chronological filters too.
```
Comment (MattDodsonEnglish) [inline content/reference/gql-types.md:91]:
```suggestion
While the `dateTime` type uses the RFC 3339 format, some string fields may use the [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) format. This depends on the object and customer requirement. For these fields, the string filters work as chronological filters too.
```

**Pull request merged**